File: MacTemplet/Cagtegory/NSTableColumn+Helper.m



//
//  NSTableColumn+Helper.m
//  MacTemplet
//
//  Created by Bin Shang on 2019/6/11.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

#import "NSTableColumn+Helper.h"

@implementation NSTableColumn (Helper)

+ (instancetype)createWithIdentifier:(NSUserInterfaceItemIdentifier)identifier title:(NSString *)title{
    NSTableColumn *column = [[self alloc] initWithIdentifier:identifier];
    column.title = title;
    column.minWidth = 40;
    column.maxWidth = CGFLOAT_MAX;
    column.headerToolTip = column.title;
    column.resizingMask = NSTableColumnAutoresizingMask;
    column.headerCell.alignment = NSTextAlignmentCenter;
    NSSortDescriptor *sort = [NSSortDescriptor sortDescriptorWithKey:column.identifier
                                                           ascending:false
                                                          comparator:^(id obj1, id obj2) {
                                                               if (obj1 < obj2) {
                                                                   return NSOrderedAscending;
                                                               }
                                                               if (obj1 > obj2) {
                                                                   return NSOrderedDescending;
                                                               }
                                                               return NSOrderedSame;
                                                           }];
    column.sortDescriptorPrototype = sort;
    return column;
}

@end

File: MacTemplet/MacCagtegory/NSTableColumn+Helper.swift
//
//  NSTableColumn+Helper.swift
//  MacTemplet
//
//  Created by Bin Shang on 2019/11/22.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

import Cocoa

@objc public extension NSTableColumn {

    /// 复用NSTableCellView
    static func create(identifier: String = String(describing: self), title: String) -> Self {
       
        let column = self.init(identifier: NSUserInterfaceItemIdentifier(identifier))
        column.title = title;
        column.minWidth = 40;
        column.maxWidth = CGFloat.greatestFiniteMagnitude;
        column.headerToolTip = column.title;
        column.resizingMask = .autoresizingMask;
        column.headerCell.alignment = .center;

        let sort = NSSortDescriptor(key: column.title, ascending: false, selector: #selector(NSString.localizedCompare(_:)))
        column.sortDescriptorPrototype = sort
        return column;
    }
    
}
